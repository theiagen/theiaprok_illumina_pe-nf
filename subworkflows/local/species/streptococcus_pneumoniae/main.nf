include { SEROBA           } from '../../../../modules/local/species_typing/streptococcus/seroba/main'
include { PBPTYPER         } from '../../../../modules/local/species_typing/streptococcus/pbptyper/main'
include { POPPUNK_DATABASE } from '../../../../modules/local/species_typing/streptococcus/poppunk/fetch/main'
include { POPPUNK          } from '../../../../modules/local/species_typing/streptococcus/poppunk/run/main'

workflow STREPTOCOCCUS_PNEUMONIAE_SPECIES_TYPING {

    take:
    ch_streptococcus // Channel of tuples [meta, assembly, reads, species]

    main:

    ch_versions = Channel.empty()
    ch_pbptyper_results = Channel.empty()
    ch_value_results = Channel.empty()

    // Extract reads and assembly
    ch_reads = ch_streptococcus.map { meta, assembly, reads, species -> [meta, reads] }
    ch_assembly = ch_streptococcus.map { meta, assembly, reads, species -> [meta, assembly] }

    if (params.paired_end && !params.ont_data) {
        SEROBA (
            ch_reads
        )
        ch_value_results = ch_value_results.mix(SEROBA.out.seroba_value_results)
        ch_versions = ch_versions.mix(SEROBA.out.versions)
    }
    
    PBPTYPER (
        ch_assembly,
        params.pbptyper_database ?: [],  // database - empty for default - can change
        params.pbptyper_min_percent_identity ?: 95,
        params.pbptyper_min_percent_coverage ?: 95
    )
    ch_pbptyper_results = PBPTYPER.out.pbtyper_predicted_tsv
    ch_versions = ch_versions.mix(PBPTYPER.out.versions)
    
    if (params.call_poppunk) {
        // Fetch PopPUNK GPS database
        POPPUNK_DATABASE (
            params.poppunk_gps_db_url ?: "https://gps-project.cog.sanger.ac.uk/GPS_6.tar.gz",
            params.poppunk_gps_external_clusters_url ?: "https://gps-project.cog.sanger.ac.uk/GPS_v6_external_clusters.csv"
        )
        ch_versions = ch_versions.mix(POPPUNK_DATABASE.out.versions)
        
        // Run PopPUNK with fetched database
        POPPUNK (
            ch_assembly,
            POPPUNK_DATABASE.out.database,
            POPPUNK_DATABASE.out.ext_clusters,
            POPPUNK_DATABASE.out.db_info
        )
        ch_value_results = ch_value_results.mix(POPPUNK.out.poppunk_value_results)
        ch_versions = ch_versions.mix(POPPUNK.out.versions)
    }

    emit:
    pbtyper_results = ch_pbptyper_results
    value_results = ch_value_results
    versions = ch_versions
}