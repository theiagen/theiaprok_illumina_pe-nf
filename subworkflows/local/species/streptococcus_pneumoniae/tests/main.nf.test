nextflow_workflow {
    name "Test Subworkflow STREPTOCOCCUS_PNEUMONIAE_SPECIES"
    script "../main.nf"
    workflow "STREPTOCOCCUS_PNEUMONIAE_SPECIES_TYPING"

    test("streptococcus_pneumoniae - paired_end with seroba") {
        setup {
            run("GUNZIP") {
                script "../../../../../modules/nf-core/gunzip/main.nf"
                process {
                    """
                    input[0] = Channel.of([
                        [id:'test_spneumoniae', single_end:false],
                        file('https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/013/276/905/GCF_013276905.1_ASM1327690v1/GCF_013276905.1_ASM1327690v1_genomic.fna.gz', checkIfExists: true)
                    ])
                    """
                }
            }
        }
        
        when {
            params {
                paired_end = true
                ont_data = false
                call_poppunk = false
            }
            workflow {
                """
                input[0] = GUNZIP.out.gunzip
                    .map { meta, assembly -> 
                        [
                            meta, 
                            assembly, 
                            [
                                file("https://github.com/sanger-pathogens/seroba/raw/refs/heads/master/seroba/tests/data/06B_1.fq.gz", checkIfExists: true),
                                file("https://github.com/sanger-pathogens/seroba/raw/refs/heads/master/seroba/tests/data/06B_2.fq.gz", checkIfExists: true)
                            ],
                            "Streptococcus pneumoniae"
                        ]
                    }
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.seroba_results,
                    workflow.out.pbptyper_results,
                    workflow.out.poppunk_results,
                    workflow.out.versions
                ).match() }
            )
        }
    }

}