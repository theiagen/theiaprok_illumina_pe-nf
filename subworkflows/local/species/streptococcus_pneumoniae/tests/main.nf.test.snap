{
    "streptococcus_pneumoniae - paired_end with seroba": {
        "content": [
            [
                [
                    {
                        "id": "test_spneumoniae",
                        "single_end": false
                    },
                    "SEROTYPE:md5,253ce0dee2b80d881ea1049b2ef50317"
                ]
            ],
            [
                [
                    {
                        "id": "test_spneumoniae",
                        "single_end": false
                    },
                    [
                        "test_spneumoniae-1A.tblastn.tsv:md5,0df2c531d8b430ca71ad82cf459b90d8",
                        "test_spneumoniae-2B.tblastn.tsv:md5,d409db68916b1821e221f68b9e93de18",
                        "test_spneumoniae-2X.tblastn.tsv:md5,34637d0371ac3a504db220d64cefa215",
                        "test_spneumoniae.fastani.tsv:md5,0e369f10dacb0fef096ca37f33a42aa5",
                        "test_spneumoniae.tsv:md5,3d536372a2bfcf5469f30ff352cfd058"
                    ]
                ]
            ],
            [
                
            ],
            [
                "versions.yml:md5,7366efccb01fd166caf077411f5273d0",
                "versions.yml:md5,73f8f98c625d969468bdbea77d17d6de"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.5"
        },
        "timestamp": "2025-07-02T21:23:47.355194"
    }
}